pkg/apis/pingcap/v1alpha1/tidbcluster.go

```diff
@@ -128,31 +128,42 @@ func (a *componentAccessorImpl) Annotations() map[string]string {
 }
 
 func (a *componentAccessorImpl) Tolerations() []corev1.Toleration {
-	tols := a.ClusterSpec.Tolerations
-	tols = append(tols, a.ComponentSpec.Tolerations...)
+	tols := a.ComponentSpec.Tolerations
+	if len(tols) == 0 {
+		tols = a.ClusterSpec.Tolerations
+	}
 	return tols
 }
 
+// BaseTiDBSpec returns the base spec of TiDB servers
 func (tc *TidbCluster) BaseTiDBSpec() ComponentAccessor {
 	return &componentAccessorImpl{&tc.Spec, &tc.Spec.TiDB.ComponentSpec}
 }
 
+// BaseTiKVSpec returns the base spec of TiKV servers
 func (tc *TidbCluster) BaseTiKVSpec() ComponentAccessor {
 	return &componentAccessorImpl{&tc.Spec, &tc.Spec.TiKV.ComponentSpec}
 }
 
+// BasePDSpec returns the base spec of PD servers
 func (tc *TidbCluster) BasePDSpec() ComponentAccessor {
 	return &componentAccessorImpl{&tc.Spec, &tc.Spec.PD.ComponentSpec}
 }
 
-func (tc *TidbCluster) BasePumpSpec() ComponentAccessor {
-	return &componentAccessorImpl{&tc.Spec, &tc.Spec.Pump.ComponentSpec}
+// BasePumpSpec returns two results:
+// 1. the base pump spec, if exists.
+// 2. whether the base pump spec exists.
+func (tc *TidbCluster) BasePumpSpec() (ComponentAccessor, bool) {
```

Comment:
The `nil` already report the non-existence of pump spec. Is it duplicate to return a bool?

Reply:
An extra returned result is designed for enforcing the programmer to check whether the pump spec is present.

Otherwise one has to remember check whether `ComponentAccssor` is nil, which is impossible in other `Base<Component>Spec` method

Reply:
OK, got it.

Reply:
The tidb spec can also be empty, for example, I just want to deploy tikv for storage.